core/src/main/java/com/linecorp/armeria/common/HttpHeadersBase.java

```diff
@@ -0,0 +1,1214 @@
+        final int i = index(h);
+
+        remove0(h, i, normalizedName);
+        for (CharSequence v : values) {
+            requireNonNullElement(values, v);
+            add0(h, i, normalizedName, v);
+        }
+    }
+
+    final void set(CharSequence name, CharSequence... values) {
+        final AsciiString normalizedName = normalizeName(name);
+        requireNonNull(values, "values");
+
+        final int h = normalizedName.hashCode();
+        final int i = index(h);
+
+        remove0(h, i, normalizedName);
+        for (CharSequence v : values) {
+            requireNonNullElement(values, v);
+            add0(h, i, normalizedName, v);
+        }
+    }
+
+    final void set(Iterable<? extends Entry<? extends CharSequence, ? extends CharSequence>> headers) {
+        requireNonNull(headers, "headers");
+        if (headers == this) {
+            return;
+        }
+
+        for (Entry<? extends CharSequence, ? extends CharSequence> e : headers) {
+            remove(e.getKey());
+        }
+
+        if (!addFast(headers)) {
+            addSlow(headers);
+        }
+    }
+
+    public HttpHeadersBase setIfAbsent(
+            Iterable<? extends Entry<? extends CharSequence, ? extends CharSequence>> headers) {
+        requireNonNull(headers, "headers");
+        final Set<AsciiString> existingNames = names();
+        if (!setIfAbsentFast(headers, existingNames)) {
+            setIfAbsentSlow(headers, existingNames);
+        }
+        return this;
+    }
+
+    private boolean setIfAbsentFast(
+            Iterable<? extends Entry<? extends CharSequence, ? extends CharSequence>> headers,
+            Set<AsciiString> existingNames) {
+
+        if (!(headers instanceof HttpHeadersBase)) {
+            return false;
+        }
+
+        final HttpHeadersBase headersBase = (HttpHeadersBase) headers;
+        HeaderEntry e = headersBase.head.after;
+
+        while (e != headersBase.head) {
+            final AsciiString key = e.key;
+            final CharSequence value = e.value;
+            assert key != null;
+            assert value != null;
+            if (!existingNames.contains(key)) {
+                add0(e.hash, index(e.hash), key, value);
+            }
+            e = e.after;
+        }
+
+        return true;
+    }
+
+    private void setIfAbsentSlow(
+            Iterable<? extends Entry<? extends CharSequence, ? extends CharSequence>> headers,
+            Set<AsciiString> existingNames) {
+
+        for (Entry<? extends CharSequence, ? extends CharSequence> header : headers) {
+            final AsciiString key = AsciiString.of(header.getKey());
+            if (!existingNames.contains(key)) {
+                add(key, header.getValue());
+            }
+        }
+    }
+
+    final void setObject(CharSequence name, Object value) {
+        requireNonNull(value, "value");
+        set(name, fromObject(value));
+    }
+
+    final void setObject(CharSequence name, Iterable<?> values) {
```

Comment:
Similar to the above comment, personal preference would be to use a different method name like `All` similar to the netty API to make the overload more precise.

Reply:
I personally like the current names because the parameter types are clearly different from each other - `CharSequence` vs `CharSequence...` vs `Iterable<CharSequence>` vs a single parameter `Iterable<Entry>`, but I'd like to make sure if there's a better option.

It's fairly easy for `setObject` because we can rename them to `setObjects`, but I have no idea about what to do about `set` because `sets()` doesn't make sense and neither `setAll()` nor `setAllObjects()` seem to explain what it does exactly. Any suggestions for better and consistent names?

Reply:
`setAllObjects` is pretty clear to me, especially making clear that the value is iterated on. The semantics are pretty different between `setObject`, which converts the object to a string, and the iteration ones, which doesn't convert the value to a string but actually iterates on it.

Reply:
I was actually asking about these methods:

- `set(Iterable<Map.Entry<>>)` and `setObject(Iterable<Map.Entry<>>)`
- `setIfAbsent(Iterable<Map.Entry<>>)`
- `add(Iterable<Map.Entry<>>)` and `addObject(Iterable<Map.Entry<>>)`

Do you think we can leave them as they are and just rename:

- `set/add(CharSequence name, String... values)` to `set/addAll()`
- `set/addObject(CharSequence name, Object... values)` to `set/addAllObjects()`

..?

Reply:
Yeah that seems fine

Reply:
I tried renaming but I somehow don't like it.. I'd like to keep it as it is unless you are strong about this one. :pray: